src/NativeBridge/PythonInterop.cpp

```diff
@@ -13,43 +20,189 @@ PythonObjectBase::~PythonObjectBase()
+        dict[name] = np::from_data(
+            data,
+            np::dtype::get_builtin<bool>(),
+            bp::make_tuple(_pData->size()),
+            bp::make_tuple(sizeof(bool)), bp::object(h));
+    }
+    break;
+    case DataKind::I1:
+    case DataKind::I2:
+    case DataKind::I4:
+    {
+        bp::handle<> h(::PyCapsule_New((void*)this, NULL, (PyCapsule_Destructor)&destroyManagerCObject));
+        np::ndarray npdata = np::from_data(
+            data,
+            np::dtype::get_builtin<T>(),
+            bp::make_tuple(_pData->size()),
+            bp::make_tuple(sizeof(float)), bp::object(h));
+        if (keyNames == nullptr)
+        {
+            dict[name] = npdata;
+        }
+        else
+        {
+            dict[name] = bp::dict();
+            dict[name]["..Data"] = npdata;
+            bp::list list;
+            for (int j = 0; j < keyNames->size(); j++)
+            {
+                bp::object obj;
+                const std::string& value = keyNames->at(j);
+                if (!value.empty())
+                {
+                    obj = bp::object(value);
+                }
+                list.append(obj);
+            }
+            dict[name]["..KeyValues"] = list;
+        }
+    }
+    break;
+    case DataKind::I8:
+    case DataKind::U1:
+    case DataKind::U2:
+    case DataKind::U4:
+    case DataKind::U8:
+    case DataKind::R4:
+    case DataKind::R8:
+    {
+        bp::handle<> h(::PyCapsule_New((void*)this, NULL, (PyCapsule_Destructor)&destroyManagerCObject));
+        dict[name] = np::from_data(
+            data,
+            np::dtype::get_builtin<T>(),
+            bp::make_tuple(_pData->size()),
+            bp::make_tuple(sizeof(T)), bp::object(h));
+    }
+    break;
+    }
 }
 
-template <class T> PythonObjectBase* PythonObjectBase::CreateObject(const int& kind, size_t nRows, size_t nColumns)
+template <>
+void PythonObjectSingle<std::string>::AddToDict(bp::dict& dict, const std::string& name, const std::vector<std::string>* keyNames)
 {
-    return new PythonObject<T>(kind, nRows, nColumns);
+    bp::list list;
+    for (size_t i = 0; i < _pData->size(); i++)
+    {
+        bp::object obj;
+        const std::string& value = _pData->at(i);
+        if (!value.empty())
+        {
+            obj = bp::object(value);
+        }
+        list.append(obj);
+    }
+    dict[name] = list;
 }
 
+
+template<class T, class T2>
+inline void PythonObjectVariable<T, T2>::AddToDict(bp::dict& dict, const std::string& name, const std::vector<std::string>* keyNames)
+{
+    const size_t numRows = this->_numRows;
+    const size_t numCols = _data.size();
+
+    for (size_t i = 0; i < numCols; i++)
+    {
+        /*
+         * Make sure all the columns are the same length.
+         */
+        for (size_t j = _data[i]->size(); j < numRows; j++)
+        {
+            _data[i]->push_back(NAN);
+        }
+
+        std::string colName = name + "." + std::to_string(i);
```

Comment:
Prepend "0"'s to the column number portion of the column name so that they sort correctly.

Reply:
This has been fixed in commit bc224b94bbc49dfbdd084678b6f75ec4cd4b02f7.

Reply:
why do you need this?

---
In reply to: [324369804](https://github.com/microsoft/NimbusML/pull/267#discussion_r324369804) [](ancestors = 324369804)